content/blog/2022/2022-04-26-introduction-to-retrofit.md

```diff
@@ -0,0 +1,536 @@
+The methods that help us further process the response are:
+- isSuccessful(): Helps determine if the response HTTP status code is 2xx.
+- body(): On success, the returns the response body. In the example above, the response gets mapped to `BookDto` class.
+- errorBody(): When the service returns a failure response, this method gives us the corresponding error data.
+
+#### Asynchronous calls
+To make an asynchronous call, the `Call` interface provides the `enqueue()` method.
+The request is triggered on a separate thread and it does not block the main thread processing.
+````java
+public void getBooksAsync(String bookRequest) {
+        Call<BookDto> bookDtoCall = libraryClient.getAllBooksWithHeaders(bookRequest);
+        bookDtoCall.enqueue(new Callback<>() {
+            @Override
+            public void onResponse(Call<BookDto> call, Response<BookDto> response) {
+                if (response.isSuccessful()) {
+                    log.info("Success response : {}", response.body());
+                } else {
+                    log.info("Error response : {}", response.errorBody());
+                }
+            }
+
+            @Override
+            public void onFailure(Call<BookDto> call, Throwable throwable) {
+                log.error("Network error occured : {}", throwable.getLocalizedMessage());
+            }
+        });
+    }
+````
+We provide implementations to the methods of the `Callback` interface. The `onResponse()` handles valid HTTP responses (both success and error) and
+`onFailure()` handles network connectivity issues.
+
+We have now covered all the basic components that will help us create a working Retrofit client in a Spring Boot application.
+In the next section, we will look at mocking the endpoints defined in the Retrofit client.
+
+### Mocking an OkHttp REST Client
+For writing Unit tests, **we will use the Spring Boot Test framework in combination with Mockito and Retrofit Mock.**
+We will include the Retrofit Mock dependency with Maven:
+````xml
+    <dependency>
+    <groupId>com.squareup.retrofit2</groupId>
+    <artifactId>retrofit-mock</artifactId>
+    <version>2.5.0</version>
+    <scope>test</scope>
+    </dependency>
+````
+Gradle:
+````groovy
+    testImplementation group: 'com.squareup.retrofit2', name: 'retrofit-mock', version: '2.5.0'
+````
+
+Next, we will test the service methods. Here we will focus on mocking the Retrofit client calls.
+Consider this example:
+````java
+    @Test
+    @DisplayName("Successful getAllBooks call")
+    public void getAllBooksTest() throws Exception {
+        LibraryAuditService libraryAuditService = applicationContext.getBean(LibraryAuditService.class);
+        String booksResponse = getBooksResponse("/response/getAllBooks.json");
+        List<BookDto> bookDtoList =
+                new ObjectMapper().readValue(booksResponse, new TypeReference<>() {
+                });
+        when(libraryClient.getAllBooks("all")).thenReturn(Calls.response(bookDtoList));
```

Comment:
The piece of code is missing where the `libraryClient` is initialized as a Mockito mock (i.e. something like `final LibraryClient libraryClient = Mockito.mock(LibraryClient.class)`. Without this info, you leave the reader guessing :).

Also, there is too much going on in this code example when all you want to show is how to mock the library client (including Spring Boot in this test is too much, for example).